heka/sandbox/filters/dollars.lua

```diff
@@ -0,0 +1,268 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+-- Run tests using `lua -e "TestMode = true;" dollars.lua`
+
+--[[
+Monitor submission sizes and counts by channel and document type. This tracks
+factors relevant to budget planning. See Bug 1179751.
+
+*Example Heka Configuration*
+
+.. code-block:: ini
+
+    [PipelineBudget]
+    type = "SandboxFilter"
+    filename = "lua_filters/dollars.lua"
+    message_matcher = "Type == 'telemetry'"
+    ticker_interval = 60
+    preserve_data = false
+
+--]]
+
+require "os"
+require "table"
+if not TestMode then
+    require "lpeg"
+    require "cjson"
+
+    fx = require "fx"
+    max_per_channel = read_config("max_per_channel") or 60
+    -- YYYYMMDD<whatever>
+    bdate_grammar = lpeg.R("09") ^ 8
+end
+
+-- Size and count by day
+local sizes = {
+    submission = {},
+    build = {},
+}
+
+-- Key count by channel (for eviction purposes)
+local counters = {
+    submission = {},
+    build = {},
+}
+
+function get_build_date(buildid)
+    if not buildid then
+        return buildid
+    end
+
+    if string.len(buildid) < 8 then
+        return nil
+    end
+
+    if not lpeg.match(bdate_grammar, buildid) then
+        return nil
+    end
+
+    return string.sub(buildid, 1, 8)
+end
+
+-- If possible, evict a key older than "newdate" to make room for the new key.
+function evict(counter, newdate)
+    local evictable_keys = {}
+    local count = 0
+    for k, v in pairs(counter) do
+        count = count + 1
+        if v ~= nil and k < newdate then
```

Comment:
setting a value to `nil` is the way I remove it from the table - does that mean that `pairs` automatically skips anything with a nil value?